lib/lookup/src/lookup_buf/mod.rs

```diff
@@ -0,0 +1,399 @@
+        let mut next = peeker.next();
+        let mut maybe_next = peeker.peek();
+        while let Some(segment) = next {
+            match segment {
+                SegmentBuf::Field(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"{}."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "{}", segment)?,
+                },
+                SegmentBuf::Index(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"[{}]."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "[{}]", segment)?,
+                },
+                SegmentBuf::Coalesce(_) => match maybe_next {
+                    Some(next) if next.is_field() || next.is_coalesce() => {
+                        write!(f, r#"{}."#, segment)?
+                    }
+                    None | Some(_) => write!(f, "{}", segment)?,
+                },
+            }
+            next = peeker.next();
+            maybe_next = peeker.peek();
+        }
+        Ok(())
+    }
+}
+
+impl LookupBuf {
+    /// Creates an lookup to the root
+    pub fn root() -> Self {
+        Self {
+            segments: VecDeque::new(),
+        }
+    }
+
+    /// Get from the internal list of segments.
+    #[instrument(level = "trace")]
+    pub fn get(&mut self, index: usize) -> Option<&SegmentBuf> {
+        self.segments.get(index)
+    }
+
+    /// Push onto the internal list of segments.
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_back(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_back(segment.into());
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_back(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_back()
+    }
+
+    #[instrument(level = "trace", skip(segment))]
+    pub fn push_front(&mut self, segment: impl Into<SegmentBuf>) {
+        self.segments.push_front(segment.into())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn pop_front(&mut self) -> Option<SegmentBuf> {
+        self.segments.pop_front()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn iter(&self) -> std::collections::vec_deque::Iter<'_, SegmentBuf> {
+        self.segments.iter()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn len(&self) -> usize {
+        self.segments.len()
+    }
+
+    #[instrument(level = "trace")]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    #[instrument(level = "trace")]
+    pub fn is_root(&self) -> bool {
+        self.is_empty()
+    }
+
+    /// Raise any errors that might stem from the lookup being invalid.
+    #[instrument(level = "trace")]
+    pub fn is_valid(&self) -> Result<(), LookupError> {
+        Ok(())
+    }
+
+    #[instrument(level = "trace")]
+    pub fn clone_lookup(&self) -> Lookup {
```

Comment:
A quick search doesn't reveal any location using this method, should we remove this?

Reply:
This is explained in the docs above: https://github.com/timberio/vector/blob/dcb257d141c66cc4033baf65149fcd647ef99aba/lib/lookup/src/lookup_buf/mod.rs#L67-L68

Perhaps there is a better name for it? I'm a bit hesitant to remove things just because we aren't using them yet as I'm sure there was a reason why they were added. This function is used in a test for the file source:

https://github.com/timberio/vector/blob/new-accessors/src/sources/file.rs#L927-L930

To be added later..

Reply:
Ah, the phrasing "shed ownership" threw me. I think "creating a reference view" or a borrow would be more customary Rust terminology. I would suggest `to_lookup`.